adapters/dmx/dmx.go

```diff
@@ -0,0 +1,297 @@
+package dmx
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+type DmxAdapter struct {
+	endpoint string
+}
+
+func NewDmxBidder(endpoint string) *DmxAdapter {
+	return &DmxAdapter{endpoint: endpoint}
+}
+
+type dmxExt struct {
+	Bidder dmxParams `json:"bidder"`
+}
+
+type dmxParams struct {
+	TagId       string `json:"tagid,omitempty"`
+	DmxId       string `json:"dmxid,omitempty"`
+	MemberId    string `json:"memberid,omitempty"`
+	PublisherId string `json:"publisher_id,omitempty"`
+	SellerId    string `json:"seller_id,omitempty"`
+}
+
+func UserSellerOrPubId(str1, str2 string) string {
+	if str1 != "" {
+		return str1
+	}
+	return str2
+}
+
+type dmxUserExt struct {
+	Eids      []dmxEids    `json:"eids,omitempty"`
+	Digitrust dmxDigitrust `json:"digitrust,omitempty"`
+}
+type dmxEids struct {
+	Source string `json:"source,omitempty"`
+	Uids   []struct {
+		ID  string `json:"id,omitempty"`
+		Ext json.RawMessage
+	} `json:"uids,omitempty"`
+}
+
+type dmxDigitrust struct {
+	ID   string `json:"id,omitempty"`
+	Keyv int    `json:"keyv,omitempty"`
+}
+
+func (adapter *DmxAdapter) MakeRequests(request *openrtb.BidRequest, req *adapters.ExtraRequestInfo) (reqsBidder []*adapters.RequestData, errs []error) {
+	var imps []openrtb.Imp
+	var rootExtInfo dmxExt
+	var publisherId string
+	var sellerId string
+	var user_ext dmxUserExt
+	var hasId []bool
+
+	var dmxReq = request
+
+	if dmxReq.User == nil {
+		if dmxReq.App == nil {
+			return nil, []error{errors.New("no user Id found and AppID, no request to DMX")}
+		}
+	}
+
+	if len(request.Imp) >= 1 {
+		err := json.Unmarshal(request.Imp[0].Ext, &rootExtInfo)
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			publisherId = UserSellerOrPubId(rootExtInfo.Bidder.PublisherId, rootExtInfo.Bidder.MemberId)
+			sellerId = rootExtInfo.Bidder.SellerId
+		}
+	}
+
+	if dmxReq.App != nil {
+		dmxReq.Site = nil
+		if dmxReq.App.ID != "" {
+			hasId = append(hasId, true)
+		}
+		if dmxReq.App.Publisher != nil {
+			dmxReq.App.Publisher.ID = publisherId
```

Comment:
The App object is shared memory among other adapters. If you want to modify it, you'll need to create a copy first. Something like...

```
appCopy := *bidRequest.App
bidRequest.App = &appCopy
bidRequest.App.Publisher = nil
```

So.. things like this is fine:
```
dmxReq.Site = nil
```
and this is not (without making a copy first)
```
dmxReq.Site.Publisher = &openrtb.Publisher{ID: publisherId}
```

In this PR, this applies to dmxReq.App, dmxReq.Site, and dmxReq.User.

Reply:
Something like this would work 

```go
if request.App != nil {
	appCopy := *request.App
	dmxReq.App = &appCopy
} else {
	dmxReq.App = nil
}
```